File: mirrors/repos/Lubu-Labs@langchain-agent-skills/skills/langsmith-trace-analyzer/scripts/download_traces.ts
#!/usr/bin/env ts-node
/**
 * Download traces from LangSmith with flexible filtering and organization.
 *
 * Features:
 * - Query by trace ID, job_id, metadata, status, or time range
 * - Organize by outcome (passed/failed/error) automatically
 * - Create manifest with trace metadata
 * - Resume interrupted downloads
 *
 * Usage:
 *   # Download specific traces
 *   ts-node download_traces.ts --trace-ids id1 id2 id3 --output ./traces
 *
 *   # Download by metadata filter
 *   ts-node download_traces.ts --project my-project --filter 'job_id=abc123' --output ./traces
 *
 *   # Download recent traces
 *   ts-node download_traces.ts --project my-project --last-hours 24 --limit 50 --output ./traces
 */

import { Client } from "langsmith";
import { spawn } from "child_process";
import * as fs from "fs/promises";
import * as path from "path";

interface TraceInfo {
  trace_id: string;
  status?: string;
  metadata?: Record<string, any>;
  start_time?: string;
  end_time?: string;
}

interface DownloadResults {
  success: string[];
  failed: Array<{ trace_id: string; error: string }>;
  by_category: Record<string, number>;
}

async function ensureDirectory(dir: string): Promise<void> {
  await fs.mkdir(dir, { recursive: true });
}

async function queryTraces(
  projectName: string,
  filterQuery?: string,
  status?: string,
  sinceHours?: number,
  limit?: number
): Promise<TraceInfo[]> {
  const client = new Client();

  const filters: string[] = [];
  if (filterQuery) filters.push(filterQuery);
  if (status) filters.push(`eq(status, "${status}")`);

  const filterStr = filters.length > 1 ? `and(${filters.join(", ")})` : filters[0];

  const opts: any = {
    projectName,
    isRoot: true,
  };

  if (filterStr) opts.filter = filterStr;
  if (sinceHours !== undefined && sinceHours !== null) {
    const startTime = new Date(Date.now() - sinceHours * 60 * 60 * 1000);
    opts.startTime = startTime;
  }

  const traces: TraceInfo[] = [];
  for await (const run of client.listRuns(opts)) {
    traces.push({
      trace_id: String(run.id),
      status: run.status,
      metadata: run.metadata || {},
      start_time: run.start_time?.toISOString(),
      end_time: run.end_time?.toISOString(),
    });

    if (limit && traces.length >= limit) break;
  }

  return traces;
}

async function fetchSingleTrace(
  traceId: string,
  outputPath: string,
  retries = 3
): Promise<[boolean, string | null]> {
  await ensureDirectory(path.dirname(outputPath));
  let lastError: string | null = null;
  const primaryArgs = [
    "trace",
    traceId,
    "--format",
    "raw",
    "--include-metadata",
    "--file",
    outputPath,
  ];
  const fallbackArgs = [
    "trace",
    traceId,
    "--format",
    "raw",
    "--file",
    outputPath,
  ];

  const runFetch = async (args: string[]): Promise<void> =>
    new Promise<void>((resolve, reject) => {
      const proc = spawn("langsmith-fetch", args);

      let stderr = "";
      proc.stderr?.on("data", (data) => {
        stderr += data.toString();
      });

      const timer = setTimeout(() => {
        proc.kill();
        reject(new Error("Timeout"));
      }, 60000);

      proc.on("close", (code) => {
        clearTimeout(timer);
        if (code === 0) resolve();
        else reject(new Error(stderr || "Unknown error"));
      });
    });

  for (let attempt = 0; attempt < retries; attempt++) {
    try {
      try {
        await runFetch(primaryArgs);
      } catch (error: any) {
        const message = error?.message || "";
        const includeFlagUnsupported =
          message.includes("include-metadata") &&
          (message.includes("No such option") || message.toLowerCase().includes("unknown option"));

        if (!includeFlagUnsupported) {
          throw error;
        }

        // Backward/forward compatibility across langsmith-fetch CLI versions.
        await runFetch(fallbackArgs);
      }

      // Verify JSON
      const content = await fs.readFile(outputPath, "utf-8");
      JSON.parse(content);

      return [true, null];
    } catch (error: any) {
      lastError = error?.message || String(error) || "Unknown error";
      if (attempt < retries - 1) {
        await new Promise((resolve) => setTimeout(resolve, 2 ** attempt * 1000));
      }
    }
  }

  return [false, lastError || "Failed after retries"];
}

function extractTraceMetadata(traceData: any): Record<string, any> {
  if (traceData?.metadata && typeof traceData.metadata === "object") return traceData.metadata;
  if (traceData?.extra?.metadata && typeof traceData.extra.metadata === "object") return traceData.extra.metadata;
  return {};
}

function extractTraceMessages(traceData: any): Array<Record<string, any>> {
  if (Array.isArray(traceData?.messages)) return traceData.messages.filter((m: any) => m && typeof m === "object");
  if (Array.isArray(traceData?.inputs?.messages)) {
    return traceData.inputs.messages.filter((m: any) => m && typeof m === "object");
  }
  return [];
}

function categorizeTrace(traceData: any): [string, string | null] {
  const metadata = extractTraceMetadata(traceData);
  const status = String(traceData?.status || metadata?.status || (traceData?.error ? "error" : "unknown"));

  if (status === "success") {
    const customMeta = metadata.custom_metadata || {};
    const reward = customMeta.reward ?? metadata.reward;
    if (reward === 1 || reward === 1.0 || reward === true) return ["passed", null];
    if (reward === 0 || reward === 0.0 || reward === false) return ["failed", null];
    return ["passed", null];
  } else if (status === "error") {
    const messages = extractTraceMessages(traceData);
    const errorBlob = String(traceData?.error || "");
    let errorType = "unknown";
    if (errorBlob.includes("GraphRecursionError")) return ["error", "GraphRecursionError"];
    if (errorBlob.includes("TimeoutError") || errorBlob.includes("AgentTimeoutError")) return ["error", "TimeoutError"];
    if (errorBlob.includes("DaytonaError")) return ["error", "DaytonaError"];

    for (let i = messages.length - 1; i >= 0; i--) {
      const content = String(messages[i].content || "");
      if (content.includes("GraphRecursionError")) {
        errorType = "GraphRecursionError";
        break;
      } else if (content.includes("TimeoutError") || content.includes("AgentTimeoutError")) {
        errorType = "TimeoutError";
        break;
      } else if (content.includes("DaytonaError")) {
        errorType = "DaytonaError";
        break;
      }
    }

    return ["error", errorType];
  }

  return ["unknown", null];
}

async function fileExists(filePath: string): Promise<boolean> {
  try {
    await fs.access(filePath);
    return true;
  } catch {
    return false;
  }
}

async function findExistingTracePath(outputDir: string, traceId: string): Promise<string | null> {
  const directCandidates = [
    path.join(outputDir, "temp", `${traceId}.json`),
    path.join(outputDir, `${traceId}.json`),
  ];

  for (const candidate of directCandidates) {
    if (await fileExists(candidate)) return candidate;
  }

  const byOutcomeDir = path.join(outputDir, "by-outcome");
  if (!(await fileExists(byOutcomeDir))) return null;

  const categories = await fs.readdir(byOutcomeDir, { withFileTypes: true });
  for (const category of categories) {
    if (!category.isDirectory()) continue;
    const categoryDir = path.join(byOutcomeDir, category.name);
    const directFile = path.join(categoryDir, `${traceId}.json`);
    if (await fileExists(directFile)) return directFile;

    const subEntries = await fs.readdir(categoryDir, { withFileTypes: true });
    for (const subEntry of subEntries) {
      if (!subEntry.isDirectory()) continue;
      const subFile = path.join(categoryDir, subEntry.name, `${traceId}.json`);
      if (await fileExists(subFile)) return subFile;
    }
  }

  return null;
}

async function main() {
  const args = process.argv.slice(2);
  const config: any = {
    project: null,
    traceIds: [],
    filter: null,
    status: null,
    lastHours: null,
    limit: null,
    output: "./langsmith-traces",
    delay: 1000,
    organize: true,
  };

  if (args.includes("--help") || args.includes("-h")) {
    console.log(`Usage:
  ts-node download_traces.ts --trace-ids <id...> [--output <dir>]
  ts-node download_traces.ts --project <name> [--filter <query>] [--status success|error]
    [--last-hours <n>] [--limit <n>] [--output <dir>] [--delay <seconds>] [--no-organize]`);
    return 0;
  }

  // Parse command line args
  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    if (arg === "--project") config.project = args[++i];
    else if (arg === "--trace-ids") {
      while (i + 1 < args.length && !args[i + 1].startsWith("--")) {
        config.traceIds.push(args[++i]);
      }
    } else if (arg === "--filter") config.filter = args[++i];
    else if (arg === "--status") config.status = args[++i];
    else if (arg === "--last-hours") config.lastHours = parseInt(args[++i]);
    else if (arg === "--limit") config.limit = parseInt(args[++i]);
    else if (arg === "--output") config.output = args[++i];
    else if (arg === "--delay") config.delay = parseFloat(args[++i]) * 1000;
    else if (arg === "--organize") config.organize = true;
    else if (arg === "--no-organize") config.organize = false;
  }

  if (config.status && !["success", "error"].includes(config.status)) {
    console.error("Error: --status must be one of: success, error");
    process.exit(1);
  }

  await ensureDirectory(config.output);

  // Get traces
  let traces: TraceInfo[];
  if (config.traceIds.length > 0) {
    traces = config.traceIds.map((id: string) => ({ trace_id: id }));
    console.log(`Downloading ${traces.length} specified traces...`);
  } else {
    if (!config.project) {
      console.error("Error: --project required when not using --trace-ids");
      process.exit(1);
    }

    // Parse filter
    let filterQuery = config.filter;
    if (filterQuery && filterQuery.includes("=")) {
      const [key, value] = filterQuery.split("=", 2);
      filterQuery = `and(eq(metadata_key, "${key}"), eq(metadata_value, "${value}"))`;
    }

    console.log(`Querying traces from project: ${config.project}`);
    traces = await queryTraces(
      config.project,
      filterQuery,
      config.status,
      config.lastHours,
      config.limit
    );
    console.log(`Found ${traces.length} matching traces`);
  }

  if (traces.length === 0) {
    console.log("No traces to download");
    return 0;
  }

  const results: DownloadResults = {
    success: [],
    failed: [],
    by_category: {},
  };

  for (let i = 0; i < traces.length; i++) {
    const trace = traces[i];
    const traceId = trace.trace_id;

    // Determine output path
    const tempPath = config.organize
      ? path.join(config.output, "temp", `${traceId}.json`)
      : path.join(config.output, `${traceId}.json`);

    const existing = await findExistingTracePath(config.output, traceId);
    if (existing) {
      console.log(
        `[${i + 1}/${traces.length}] SKIP: ${traceId} (already exists at ${path.relative(config.output, existing)})`
      );
      results.success.push(traceId);
      continue;
    }

    console.log(`[${i + 1}/${traces.length}] Downloading: ${traceId}`);

    const [success, error] = await fetchSingleTrace(traceId, tempPath);

    if (success) {
      // Categorize and move if organizing
      if (config.organize) {
        const content = await fs.readFile(tempPath, "utf-8");
        const traceData = JSON.parse(content);

        const [category, subcategory] = categorizeTrace(traceData);
        results.by_category[category] = (results.by_category[category] || 0) + 1;

        const finalDir = subcategory
          ? path.join(config.output, "by-outcome", category, subcategory)
          : path.join(config.output, "by-outcome", category);

        await ensureDirectory(finalDir);
        const finalPath = path.join(finalDir, `${traceId}.json`);

        await fs.rename(tempPath, finalPath);
        console.log(`  -> ${path.relative(config.output, finalPath)}`);
      } else {
        console.log(`  -> ${path.relative(config.output, tempPath)}`);
      }

      results.success.push(traceId);
    } else {
      console.log(`  FAILED: ${error}`);
      results.failed.push({ trace_id: traceId, error: error || "Unknown" });
    }

    // Rate limiting
    if (i < traces.length - 1) {
      await new Promise((resolve) => setTimeout(resolve, config.delay));
    }
  }

  // Create manifest
  const manifest = {
    created_at: new Date().toISOString(),
    project: config.project,
    filter: config.filter,
    total_traces: traces.length,
    downloaded: results.success.length,
    failed: results.failed.length,
    by_category: results.by_category,
  };

  const manifestPath = path.join(config.output, "manifest.json");
  await fs.writeFile(manifestPath, JSON.stringify(manifest, null, 2));

  console.log("\n" + "=".repeat(60));
  console.log("DOWNLOAD COMPLETE");
  console.log("=".repeat(60));
  console.log(`Success: ${results.success.length}`);
  console.log(`Failed: ${results.failed.length}`);

  if (Object.keys(results.by_category).length > 0) {
    console.log("\nBy Category:");
    Object.entries(results.by_category)
      .sort()
      .forEach(([cat, count]) => {
        console.log(`  ${cat}: ${count}`);
      });
  }

  if (results.failed.length > 0) {
    console.log("\nFailed downloads:");
    results.failed.slice(0, 10).forEach((item) => {
      console.log(`  - ${item.trace_id}: ${item.error}`);
    });
  }

  console.log(`\nManifest written to: ${manifestPath}`);

  return results.failed.length > 0 ? 1 : 0;
}

main()
  .then((code) => process.exit(code))
  .catch((error) => {
    console.error(error instanceof Error ? error.message : String(error));
    process.exit(1);
  });

File: mirrors/repos/soba-labs@langchain-agent-skills/skills/langsmith-trace-analyzer/scripts/download_traces.py
#!/usr/bin/env python3
"""
Download traces from LangSmith with flexible filtering and organization.

Features:
- Query by trace ID, job_id, metadata, status, or time range
- Organize by outcome (passed/failed/error) automatically
- Create manifest with trace metadata
- Resume interrupted downloads

Usage:
    # Download specific traces by ID
    uv run skills/langsmith-trace-analyzer/scripts/download_traces.py --trace-ids id1 id2 id3 --output ./traces

    # Download by job_id metadata
    uv run skills/langsmith-trace-analyzer/scripts/download_traces.py --project my-project --filter 'job_id=abc123' --output ./traces

    # Download recent traces
    uv run skills/langsmith-trace-analyzer/scripts/download_traces.py --project my-project --last-hours 24 --limit 50 --output ./traces

    # Download error traces only
    uv run skills/langsmith-trace-analyzer/scripts/download_traces.py --project my-project --status error --output ./traces
"""

import argparse
import json
import subprocess
import sys
import time
from collections import defaultdict
from datetime import datetime, timedelta, timezone
from pathlib import Path
from typing import Any, Dict, List, Optional


def ensure_langsmith_fetch():
    """Check if langsmith-fetch is installed."""
    try:
        subprocess.run(["langsmith-fetch", "--help"], capture_output=True, check=True)
    except (subprocess.CalledProcessError, FileNotFoundError):
        print("Error: langsmith-fetch not installed. Install with:")
        print("  uv pip install langsmith-fetch")
        sys.exit(1)


def ensure_langsmith_sdk():
    """Check if langsmith SDK is installed."""
    try:
        import langsmith
        return True
    except ImportError:
        print("Error: langsmith SDK not installed. Install with:")
        print("  uv pip install langsmith")
        sys.exit(1)


def query_traces(
    project_name: str,
    filter_query: Optional[str] = None,
    status: Optional[str] = None,
    since_hours: Optional[int] = None,
    limit: Optional[int] = None
) -> List[Dict]:
    """Query LangSmith for traces matching criteria."""
    from langsmith import Client

    client = Client()
    filters = []

    if filter_query:
        filters.append(filter_query)

    if status:
        filters.append(f'eq(status, "{status}")')

    filter_str = f'and({", ".join(filters)})' if len(filters) > 1 else filters[0] if filters else None

    kwargs = {
        "project_name": project_name,
        "is_root": True,
    }

    if filter_str:
        kwargs["filter"] = filter_str

    if since_hours is not None:
        kwargs["start_time"] = datetime.now(timezone.utc) - timedelta(hours=since_hours)

    traces = []
    for run in client.list_runs(**kwargs):
        traces.append({
            "trace_id": str(run.id),
            "status": run.status,
            "metadata": run.metadata or {},
            "start_time": run.start_time.isoformat() if run.start_time else None,
            "end_time": run.end_time.isoformat() if run.end_time else None,
        })

        if limit and len(traces) >= limit:
            break

    return traces


def fetch_single_trace(trace_id: str, output_path: Path, retries: int = 3) -> tuple[bool, Optional[str]]:
    """Fetch a single trace using langsmith-fetch CLI."""
    output_path.parent.mkdir(parents=True, exist_ok=True)

    cmd = [
        "langsmith-fetch", "trace", trace_id,
        "--format", "raw",
        "--include-metadata",
        "--file", str(output_path)
    ]
    fallback_cmd = [
        "langsmith-fetch", "trace", trace_id,
        "--format", "raw",
        "--file", str(output_path)
    ]

    for attempt in range(retries):
        try:
            result = subprocess.run(cmd, capture_output=True, text=True, timeout=60)
            if result.returncode != 0 and (
                "include-metadata" in (result.stderr or "")
                and ("No such option" in (result.stderr or "") or "unknown option" in (result.stderr or "").lower())
            ):
                # Backward/forward compatibility across langsmith-fetch CLI versions.
                result = subprocess.run(fallback_cmd, capture_output=True, text=True, timeout=60)

            if result.returncode == 0 and output_path.exists():
                # Verify valid JSON
                with open(output_path) as f:
                    json.load(f)
                return True, None

            error = result.stderr or "Unknown error"

        except subprocess.TimeoutExpired:
            error = "Timeout"
        except json.JSONDecodeError as e:
            error = f"Invalid JSON: {e}"
        except Exception as e:
            error = str(e)

        if attempt < retries - 1:
            time.sleep(2 ** attempt)

    return False, error


def _extract_trace_metadata(trace_data: Dict[str, Any]) -> Dict[str, Any]:
    """Extract metadata from supported trace payload shapes."""
    metadata = trace_data.get("metadata")
    if isinstance(metadata, dict):
        return metadata

    extra = trace_data.get("extra")
    if isinstance(extra, dict):
        extra_metadata = extra.get("metadata")
        if isinstance(extra_metadata, dict):
            return extra_metadata

    return {}


def _extract_trace_messages(trace_data: Dict[str, Any]) -> List[Dict[str, Any]]:
    """Extract messages from trace payload."""
    messages = trace_data.get("messages")
    if isinstance(messages, list):
        return [m for m in messages if isinstance(m, dict)]

    inputs = trace_data.get("inputs")
    if isinstance(inputs, dict):
        input_messages = inputs.get("messages")
        if isinstance(input_messages, list):
            return [m for m in input_messages if isinstance(m, dict)]

    return []


def categorize_trace(trace_data: Dict[str, Any]) -> tuple[str, Optional[str]]:
    """Categorize trace into passed/failed/error with subcategory."""
    metadata = _extract_trace_metadata(trace_data)
    status = trace_data.get("status") or metadata.get("status")
    if not status and trace_data.get("error"):
        status = "error"
    status = str(status or "unknown")

    if status == "success":
        # Check for task outcome if available
        custom_meta = metadata.get("custom_metadata", {}) if isinstance(metadata.get("custom_metadata"), dict) else {}
        reward = custom_meta.get("reward", metadata.get("reward"))
        if reward in (1, 1.0, True):
            return "passed", None
        if reward in (0, 0.0, False):
            return "failed", None
        return "passed", None

    if status == "error":
        # Try to extract error type from error field and messages.
        messages = _extract_trace_messages(trace_data)
        error_blob = str(trace_data.get("error", ""))
        if "GraphRecursionError" in error_blob:
            return "error", "GraphRecursionError"
        if "TimeoutError" in error_blob or "AgentTimeoutError" in error_blob:
            return "error", "TimeoutError"
        if "DaytonaError" in error_blob:
            return "error", "DaytonaError"

        error_type = "unknown"
        for msg in reversed(messages):
            content = str(msg.get("content", ""))
            if "GraphRecursionError" in content:
                error_type = "GraphRecursionError"
                break
            if "TimeoutError" in content or "AgentTimeoutError" in content:
                error_type = "TimeoutError"
                break
            if "DaytonaError" in content:
                error_type = "DaytonaError"
                break

        return "error", error_type

    return "unknown", None


def find_existing_trace_path(output_dir: Path, trace_id: str) -> Optional[Path]:
    """Locate an already-downloaded trace by id."""
    candidates = [
        output_dir / "temp" / f"{trace_id}.json",
        output_dir / f"{trace_id}.json",
    ]

    for candidate in candidates:
        if candidate.exists():
            return candidate

    by_outcome_dir = output_dir / "by-outcome"
    if by_outcome_dir.exists():
        for path in by_outcome_dir.rglob(f"{trace_id}.json"):
            return path

    return None


def main():
    parser = argparse.ArgumentParser(
        description="Download traces from LangSmith with flexible filtering",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog=__doc__
    )

    parser.add_argument("--project", help="LangSmith project name")
    parser.add_argument("--trace-ids", nargs="+", help="Specific trace IDs to download")
    parser.add_argument("--filter", help="Custom filter query (e.g., 'job_id=abc123')")
    parser.add_argument("--status", choices=["success", "error"], help="Filter by status")
    parser.add_argument("--last-hours", type=int, help="Download traces from last N hours")
    parser.add_argument("--limit", type=int, help="Maximum traces to download")
    parser.add_argument("--output", default="./langsmith-traces", help="Output directory")
    parser.add_argument("--delay", type=float, default=1.0, help="Delay between requests (seconds)")
    parser.add_argument(
        "--organize",
        action=argparse.BooleanOptionalAction,
        default=True,
        help="Organize by outcome",
    )

    args = parser.parse_args()

    ensure_langsmith_fetch()

    output_dir = Path(args.output)
    output_dir.mkdir(parents=True, exist_ok=True)

    # Get trace IDs
    if args.trace_ids:
        traces = [{"trace_id": tid, "status": "unknown", "metadata": {}} for tid in args.trace_ids]
        print(f"Downloading {len(traces)} specified traces...")
    else:
        if not args.project:
            print("Error: --project required when not using --trace-ids")
            sys.exit(1)

        ensure_langsmith_sdk()

        # Parse filter if provided
        filter_query = None
        if args.filter:
            if "=" in args.filter:
                key, value = args.filter.split("=", 1)
                filter_query = f'and(eq(metadata_key, "{key}"), eq(metadata_value, "{value}"))'
            else:
                filter_query = args.filter

        print(f"Querying traces from project: {args.project}")
        traces = query_traces(
            args.project,
            filter_query=filter_query,
            status=args.status,
            since_hours=args.last_hours,
            limit=args.limit
        )
        print(f"Found {len(traces)} matching traces")

    if not traces:
        print("No traces to download")
        return 0

    # Download traces
    results = {"success": [], "failed": [], "by_category": defaultdict(int)}

    for i, trace_info in enumerate(traces, 1):
        trace_id = trace_info["trace_id"]

        temp_path = output_dir / "temp" / f"{trace_id}.json" if args.organize else output_dir / f"{trace_id}.json"

        # Skip if already exists anywhere in the output directory structure.
        existing = find_existing_trace_path(output_dir, trace_id)
        if existing is not None:
            relative = existing.relative_to(output_dir)
            print(f"[{i}/{len(traces)}] SKIP: {trace_id} (already exists at {relative})")
            results["success"].append(trace_id)
            continue

        print(f"[{i}/{len(traces)}] Downloading: {trace_id}")

        success, error = fetch_single_trace(trace_id, temp_path)

        if success:
            # Categorize and move to proper location if organizing
            if args.organize:
                with open(temp_path) as f:
                    trace_data = json.load(f)

                category, subcategory = categorize_trace(trace_data)
                results["by_category"][category] += 1

                if subcategory:
                    final_dir = output_dir / "by-outcome" / category / subcategory
                else:
                    final_dir = output_dir / "by-outcome" / category

                final_dir.mkdir(parents=True, exist_ok=True)
                final_path = final_dir / f"{trace_id}.json"

                temp_path.rename(final_path)
                print(f"  -> {final_path.relative_to(output_dir)}")
            else:
                print(f"  -> {temp_path.relative_to(output_dir)}")

            results["success"].append(trace_id)
        else:
            print(f"  FAILED: {error}")
            results["failed"].append({"trace_id": trace_id, "error": error})

        # Rate limiting
        if i < len(traces):
            time.sleep(args.delay)

    # Create manifest
    manifest = {
        "created_at": datetime.now(timezone.utc).isoformat(),
        "project": args.project,
        "filter": args.filter,
        "total_traces": len(traces),
        "downloaded": len(results["success"]),
        "failed": len(results["failed"]),
        "by_category": dict(results["by_category"]),
    }

    manifest_path = output_dir / "manifest.json"
    with open(manifest_path, "w") as f:
        json.dump(manifest, f, indent=2)

    print(f"\n{'=' * 60}")
    print("DOWNLOAD COMPLETE")
    print(f"{'=' * 60}")
    print(f"Success: {len(results['success'])}")
    print(f"Failed: {len(results['failed'])}")

    if results["by_category"]:
        print("\nBy Category:")
        for cat, count in sorted(results["by_category"].items()):
            print(f"  {cat}: {count}")

    if results["failed"]:
        print("\nFailed downloads:")
        for item in results["failed"][:10]:
            print(f"  - {item['trace_id']}: {item['error']}")

    print(f"\nManifest written to: {manifest_path}")

    return 0 if not results["failed"] else 1


if __name__ == "__main__":
    sys.exit(main())
